### ISSUE TITLE
Webview app crashes with terrain 3D

### ISSUE THREAD

**Issue by HarelM** [maplibre/maplibre-gl-js#1368]
It seems that the latest memory fix was not enough to make the app stable, I think.
Related to #1291 
The app crashes after panning around sometime.

**maplibre-gl-js version**: 2.2.0-pre.3

**browser**: Android Webview inside Cordova app

I tried to make the following minimal reproduction but was not able to cause it to crash on my device:
https://angular-ivy-maplibre-terrain.stackblitz.io
https://stackblitz.com/edit/angular-ivy-maplibre-terrain?file=src/app/app.component.ts
The following is the app if anyone would like to install it:
https://ci.appveyor.com/project/IsraelHikingHost/site/builds/44123133/job/uoqpvn8y0d5d7wbe
The code for it is in a branch here:
https://github.com/IsraelHikingMap/Site/tree/3d/IsraelHiking.Web

It's an angular app wrapped in Cordova webview basically.

### Steps to Trigger Behavior

 1. Open the app
 2. Tilt the map to max
 3. Pan the map until the os kills it

### Link to Demonstration
https://angular-ivy-maplibre-terrain.stackblitz.io

### Expected Behavior
App should not crash

### Actual Behavior
App crashes on a samsung galaxy s20 device (a strong device).

@prozessor13 let me know if there's anything I can test in order to see what's the limitation of the current code - maybe set a different max cache limit or something...?

I haven't dug into the logcat device log to see that it actually crashes on out of memory, but I can do that if needed...

**Comment by prozessor13** [comment]
Hello Harel, on which Mobile-Phone?

On my IPhone 6s it crashes instant,
on my Nokia 7.2 there is no problem.

**Comment by HarelM** [comment]
Samsung Galaxy S20, running Android 12 (a "strong" phone).

**Comment by prozessor13** [comment]
Hello Harel,

i tested also on iPhone X & Google Pixel 3a, and both work without a problem. So, i can reproduce the crash only on my iPhone 6s, but because i do not have an apple-computer i cannot get any error-messages.

So, can you send me some kind of error-message from your device?

**Comment by HarelM** [comment]
The link isn't creating the crash, it only happens in the app.
I'll send a link here to aab file so you can test it. Do you have an android device to test with?

**Comment by prozessor13** [comment]
Ah, ok. yes i have an android device, but never did something with android development, so i have absolutely no idea how to start. Would be great if you can digg for an error-message of the crashed app.

**Comment by HarelM** [comment]
Ok, I'll see what I can do. Thanks!

**Comment by HarelM** [comment]
I'm able to reproduce this issue easily but I can't find anything meaningful in the logcat log I'm exporting from the device I've used to simulate this issue.
Not sure how to proceed either.
If anyone is experienced in reading logcal logs I'm attaching the log here.
[logcat.log](https://github.com/maplibre/maplibre-gl-js/files/9439285/logcat.log)
The last message from the app can be found here:
```
08-28 10:23:54.627  1567  2219 I ActivityManager: Process il.org.osm.israelhiking (pid 4730) has died: fore TOP (177,175)
08-28 10:23:54.628  1567  3205 V WindowManager: Remove Window{26be942 u0 il.org.osm.israelhiking/il.org.osm.israelhiking.MainActivity}: mSurfaceController=Surface(name=il.org.osm.israelhiking/il.org.osm.israelhiking.MainActivity[4730])/@0xbdaa3c0 mAnimatingExit=false mRemoveOnExit=false mHasSurface=true surfaceShowing=true isAnimationSet=false app-animation=false mWillReplaceWindow=false inPendingTransaction=false mDisplayFrozen=false callers=com.android.server.wm.WindowState.access$300:280 com.android.server.wm.WindowState$DeathRecipient.binderDied:3318 android.os.BinderProxy.sendDeathNotice:1209 <bottom of call stack> <bottom of call stack> <bottom of call stack> 

```
App Process ID: 4730 (Third column), can be used for filtering
App ID: il.org.osm.israelhiking

From my experience (not data oriented) the app becomes more and more sluggish until it finally crashes.
I can't think of anything else besides memory issue as was before but less accute.
If there's a branch or code change you want me to test out please let me know and I'll be happy to see if any fix helps.
If you want to be able to do these tests on your own we might be able to create a smaller app to allow easier reproduction, IDK, let me know how I can help out...

**Comment by prozessor13** [comment]
hmm. i cannot find anything useful in the logcat.log as well.
What i noticed is, that your app is also very slow in the browser.
What would be interesting. Try another stylesheet with only one layer:

```
{
   version: 8,
   sources: {
      osm: {
         type: 'raster',
         tiles: ['https://a.tile.openstreetmap.org/{z}/{x}/{y}.png'],
         tileSize: 256,
         attribution: '&copy; OpenStreetMap Contributors',
         maxzoom: 19
      },
      terrainSource: {
         type: 'raster-dem',
         url: 'https://demotiles.maplibre.org/terrain-tiles/tiles.json',
         tileSize: 256
      }
   },
   layers: [
      {
         id: 'osm',
         type: 'raster',
         source: 'osm'
     }
   ],
   terrain: {
      source: 'terrainSource',
      exaggeration: 1
   }
}
```

May the issue is because of texture-caching specifically for your stylesheet.

PS: https://angular-ivy-maplibre-terrain.stackblitz.io/ is no longer working, but it would be interesting to debug a little more in browser.

**Comment by HarelM** [comment]
Just tested https://angular-ivy-maplibre-terrain.stackblitz.io/ and it works as expected. not sure what's not working for you.
Regarding the app - so yes, I agree, it's a lot slower with the terrain is on - you can download it from the store and see that it's not super fact, but not that slow (the store version is without the terrain)
https://play.google.com/store/apps/details?id=il.org.osm.israelhiking&hl=en&gl=US
In theory, you can add a layer in that app that is pointing the raster source to see how it works without our complicated style.
I'll test that as it is worth understanding better what causes this issue, thanks for the direction! :-)

**Comment by prozessor13** [comment]
on my browser all requests from c.staticblitz.io are pending. Last week they worked without any problem.

The thing about texture-caching and stylesheet is. RenderToTexture logic will create stacks of layers which can be rendered once onto a texture. symbol & hillshading in between creates every time new stacks to get correct visibility in 3d. For example: debug-pages/terrain.html creates 2 textures for each tile. Our toursprung-terrain stylesheet creates only one texture, because in our case hillshading is added as raster-source. So may your stylesheet creates even more than 2 textures per render-to-texture tile, and may this consumes too much memory. When looking onto your test-page i could easily see how much textures your stylesheet will create.

**Comment by prozessor13** [comment]
So, your test-page is working again, and i checked your stylesheet. As thought your stylesheet produces 9 Textures per render-to-texture tile. This is because you mix render-to-texture and symbol layers to often.

Solution:
* re-structure your stylesheet, and group all render-to-texture layers https://github.com/maplibre/maplibre-gl-js/blob/main/src/render/render_to_texture.ts#L35
* or change render-logic and render always only into one texture, but this will create different rendering results, than in 2D.

To verify that the app crashes because of the stylesheet, can you please test my simple stylesheet from above?

**Comment by HarelM** [comment]
I meant to test a raster source and terrain last night but didn't got to it, I hope to be able to do it tonight.
Can you clarify what we need to do in order to restructure the style better?
What layers we need to place where and how?
In general I think the style is very complicated and heavy to render, but it's OK right now, if there are way to optimize rendering in general (i.e. 2D too) I think we can benefit from it.
A lot of the style is composed this way since the order is very important and we have a lot of info in our style.
Having said that, this is a valid use case and we need to think maybe how to properly address these kind of cases...

**Comment by prozessor13** [comment]
totally agree, this is a common usecase!

technically because of performance: all rtt (= render-to-texture) qualified layers are not rendered directly onto texture, instead the will grouped in stacks, and than rendered all by once. Every time no No-rtt layer comes a new stack will be created, which results in an extra texture, so

layers: [ fill, fill, line, hillshade, fill, fill, symbol, fill, fill, symbol ]
will result in:
* fill, fill, line -> first texture -> draped over terrain, this texture can be cached and needes rerender only on stylechange
* hillshade -> second texture -> draped over terrain, this texture needs rerender on every camera-movement
* fill, fill -> third texture -> draped over terrain, this texture can be cached and needes rerender only on stylechange
* symbol -> rendered directly onto screen
* fill, fill -> fourth texture -> draped over terrain, this texture can be cached and needes rerender only on stylechange
* symbol -> rendered directly onto screen

so an optimized stylesheet would be:
layers: [ fill, fill, line, hillshade, fill, fill, fill, fill, symbol, symbol ]
in this scenario only three textures will be nedded

to optimize more:
in our stylesheet *toursprung-terrain* hillshading is included as a raster-layer, and so only one texture is needed.

so the stack logic is needed to render the correct visibility of all layers, but this costs a lot of performance if rrt & noRtt layers mixed.

**Comment by HarelM** [comment]
I've checked with OSM raster layer only and the app does not crash (or at least I couldn't make it crash when trying).
It is slower with Terrain and seems to get just a bit slower as you pan around but stops being slower and keeps the current experience going forward. This is a feeling, I can't put it into numbers.
regardless, the optimization you mentioned is interesting to research. Thanks!
I don't know how to proceed obviously :-)

**Comment by zstadler** [comment]
> technically because of performance: all rtt (= render-to-texture) qualified layers are not rendered directly onto texture, instead the will grouped in stacks, and than rendered all by once. Every time no No-rtt layer comes a new stack will be created, which results in an extra texture ...

Which layers are considered non-rtt?
What about symbol layers that do not change with tilt and rotation?

**Comment by prozessor13** [comment]
> Which layers are considered non-rtt

https://github.com/maplibre/maplibre-gl-js/blob/main/src/render/render_to_texture.ts#L35

> What about symbol layers that do not change with tilt and rotation?

Currently they are not rendered to texture, but would be a good idea!

Thoughts to make the rendering more stable:

currently a number terrain_source_cache tiles are kept in cache. The cachsize is defined by the underlying raster-dem source (https://github.com/maplibre/maplibre-gl-js/blob/main/src/source/terrain_source_cache.ts#L53) and every of such a terrain_source_cache tile keeps their textures in cache `tile.textures`. But in case of israel-hiking-map, and because every tile has 9 textures this gets to much GPU memory and the app crashes. So i think it is necessary to set the cache to a max of textures and not a max of tiles. This should be not to difficult.

But the less textures are cached, the more bad will the performacne. So my question are:

* it looks that creating a lot of textures is very slow. In my world i think the more are in cache, the better, but it looks that performance decreeses with the number of textures. So may it is faster to rerender textures more often (as in 2d all layers rerenders every requestAnimationFrame). Tests has to to done.
* Secondly does anyone know if the GPU run render-tasks in paralell? I think so. So may i create more render-to-texture calls in paralell (with more framebuffers) and not seriell (in one framebuffer).

**Comment by HarelM** [comment]
@JannikGM do you know the answer to the above question regarding GPU parallel tasks?
I can't help much in answering your questions, but I'd be happy to test anything you would like me to test :-)

**Comment by zstadler** [comment]
> > What about symbol layers that do not change with tilt and rotation?
> 
> Currently they are not rendered to texture, but would be a good idea!

Please consider implementing that. The style has a few symbols that annotate lines. These annotations don't change with map rotation and tilt:

- Fence annotated with an x
  ![image](https://user-images.githubusercontent.com/1304610/188287070-37d18498-8472-4408-a682-b2c216723622.png)
- Cliff annotated with a brown triangle
  ![image](https://user-images.githubusercontent.com/1304610/188287132-f2b08b07-06e4-4be9-8b49-b1492a5f12df.png)
- Minefield annotated with a red triangle
  ![image](https://user-images.githubusercontent.com/1304610/188287205-53628388-4cf0-4aa3-983d-2be217005ab4.png)

The line annotation symbol layers could be grouped separately from the corresponding line layers, but that would be something the typical style designer would do.

**Comment by prozessor13** [comment]
@zstadler i started to implement this, but as always this is not so easy as expected, because symbol layout properties can have expressions, and so it is not so easy to check if symbol is stable or not. Would be great i you can make a PR for this feature. It should be not to difficult. I think changing this:
https://github.com/maplibre/maplibre-gl-js/blob/main/src/render/render_to_texture.ts#L35
lookup-variable into a function should be sufficient.

**Comment by zstadler** [comment]
@prozessor13 
I suggest that only symbol layers with  a **constant** `"map"` value for both [`icon-pitch-alignment`](https://maplibre.org/maplibre-gl-js-docs/style-spec/layers/#layout-symbol-icon-pitch-alignment)  and [`icon-rotation-alignment`](https://maplibre.org/maplibre-gl-js-docs/style-spec/layers/#layout-symbol-icon-rotation-alignment) will be rendered to texture.

**Comment by prozessor13** [comment]
I created an experimental branch with parallel render-to-texture rendering with a significant improvement in performance (see terrain-satellite demo).
https://github.com/prozessor13/maplibre-gl-js/tree/parallel_rtt
Note, this is a quick&dirty hack. This branch lacks of:
* limit parallel framebuffers/textures (this code creates as much as needed to render the current scene in once)
* removing no longer needed code from serial-rendering

But before i continue, there needs to be:
* a decision in rendering hillshade and symbols layers to reduce needed RTT textures
* i have some software-design questions, how to structure the new code. @HarelM can we make a call so that my further work goes into the right direction?

**Comment by HarelM** [comment]
Sure, I'd be happy to help with the design.

**Comment by HarelM** [comment]
As a side note, looking at the code you wrote it seems that you are creating some sort of "FBO pool" and some indexing around it for performance. I think this is probably a good direction, I would consider moving this logic to a different file/class and use it/pass it where needed.
As a general rule, whenever you pass `this` to a method your are probably doing something wrong :-)
But let's meet so that we can go over the code together.

**Comment by HarelM** [comment]
As was stated before in this thread, the problem is more severe on iOS devices, probably due to how safari has implemented stuff and also due to lower memory restrictions which I believe exists on iOS devices.
I was able to reproduce the issue more easily on iOS both an old iPad I have and also on iPhone while running my app.

**Comment by JannikGM** [comment]
> So, your test-page is working again, and i checked your stylesheet. As thought your stylesheet produces 9 Textures per render-to-texture tile. This is because you mix render-to-texture and symbol layers to often.

I'm surprised this can be a problem. If symbol layers are excluded from draping, why would they cause a new texture to be created?

I do understand that you'll want to cache hillshade and fill separately (although, maybe not practical to cache this, if you end up with too many layers to cache), but for symbols I don't understand the issue, unless they are supposed to get draped (do we have `icon-pitch-alignment` draping support?).

> it looks that creating a lot of textures is very slow. In my world i think the more are in cache, the better, but it looks that performance decreeses with the number of textures. So may it is faster to rerender textures more often (as in 2d all layers rerenders every requestAnimationFrame). Tests has to to done.

That doesn't sound right (unless you re-create them everytime or rebind them).
Unless you have so many textures that they get evicted from VRAM, you should be able to have as many textures as you want without performance loss.

> Secondly does anyone know if the GPU run render-tasks in paralell? I think so. So may i create more render-to-texture calls in paralell (with more framebuffers) and not seriell (in one framebuffer).

> @JannikGM do you know the answer to the above question regarding GPU parallel tasks?

OpenGL (~ WebGL) has no such concept because it's a high level API; it doesn't really dictate how an implementation maps to hardware.

However, OpenGL commands usually add to a list that the GPU will process in the background asynchronously; a typical GPU also has various stages which often run like a pipeline. Some changes need to flush the entire pipeline, so they should be avoided.

How drivers handle access to resources varies (= what happens if you modify a texture from the CPU side, while the GPU is still using it to render already submitted commands?).
What @prozessor13 is doing is avoiding one of the problems (blocking until the framebuffer isn't used anymore) but causes a new one (many state changes which flush the pipeline).

> I created an experimental branch with parallel render-to-texture rendering with a significant improvement in performance (see terrain-satellite demo). https://github.com/prozessor13/maplibre-gl-js/tree/parallel_rtt Note, this is a quick&dirty hack.

The FBO allocation code looks overcomplicated.

Instead of the current approach, it probably makes sense to have a ring-buffer of FBOs (= use `fbo[currentFbo++ % fboCount]`).
A ring-buffer as small as 2 FBOs could be enough.

Also, costly operations should be avoided if possible. This includes everything which changes the framebuffer dimensions / location in memory. Switching state (especially shaders) should also be avoided (= the less you interact with OpenGL, the better).

So, another option can be to allocate a larger FBO, and then render tiles into different regions of the larger FBO using a scissor test, clipping, or similar (= [generating a texture atlas](https://eu-images.contentstack.com/v3/assets/blt95b381df7c12c15d/blt762a5279295ac3f4/611e3156be258b650745d05d/figure1.jpg)).
That way, you can avoid changing the FBO altogether (avoid pipeline flush) and during rendering, you don't have to switch the source texture either. However, you'll have to respect the maximum render-target size of the hardware and caching within the GPU might be less effective + there's the additional cost of clipping.

---

Another factor for performance would be to do as much CPU (and GPU work) as possible between submitting render commands for an FBO and using the resulting texture.

**Comment by prozessor13** [comment]
After some searching (using the right keywords) if found out two things.

* switching FBOs is faster then changing its target-texture. https://stackoverflow.com/questions/6432773/opengl-using-only-one-framebuffer-and-switching-target-textures
* switching FBO and after that clearing the fbo is also a lot faster, because no memory will copied. https://stackoverflow.com/questions/10729352/framebuffer-fbo-render-to-texture-is-very-slow-using-opengl-es-2-0-on-android

Both concepts are already used in the paralell-rtt branch.

**Comment by prozessor13** [comment]
For documentation my experiences in rtt performance:

**My first approach was:**
* render layer by layer
* create one fbo for render to texture
* remember all needed textures in terrain-tiles
* check if layer is draped. if yes: 1) switch FBO from screen to rtt, 2) bind target tile-texture, 3) render next layer onto the texture

This was bad in every case because:
* switch fbo every layer without clear -> slow
* rebind texture every layer -> slow
 

**My second approach was (as in current master):**
* still one fbo for render to texture
* still remember all needed textures in terrain-tiles
* render all draped layers at once (in stacks, splitted by non draped layers)

This was better, because:
* framebuffer switch could be done much less and with the clear cmd -> good
* still texture rebinding -> bad


**Paralell-rtt branch (2.version):**
* create a pool with fbo's, each fbo has its own target texture.
* remember pool-index and not texture in tiles for caching.
* When pool is full, render intermediate result, and reuse pool objects

With this change i think the performacne is now much much better.

@JannikGM : can you agree with my experiences?

PS: i also read about bindless textures, but i think this change would be too much work!

**Comment by zstadler** [comment]
> In the new branch `hillshade` will draped to terrain combined with all other layers.

Would this branch retain the dynamic behavior of `hillshade` layers when the map is rotated in 2D?
In 2D, the [hillshade-illumination-direction](https://maplibre.org/maplibre-gl-js-docs/style-spec/layers/#paint-hillshade-hillshade-illumination-direction) is maintained relative to the viewpoint

**Comment by prozessor13** [comment]
Yes, that's the case, but that decreases the render-calls in normal stylestheets by factor 3!

There are at least 2 ways to deal with this:

* set in 3d *hillshade-illumination-direction* to *map*, as @HarelM somewhere else (i cant find it now) explained that the 'viewport' option is needed basically in 2d to get a feeling what are hills and what a valleys.
* add a logic to rerender all rtt tiles on *orientation* change-ends, so during the animation you get performant rendering and in the end the correct result.

May a general discussion should be started for this issue?

**Comment by HarelM** [comment]
Hillshading is already part of #1607. We can continue the discussion there.

**Comment by HarelM** [comment]
This should hopefully be fixed after the merge of #1671. I'll see if I can move a few more fixes in before I create a new pre-release to be able to better test this.

**Comment by HarelM** [comment]
This is not fixed. Thanks @prozessor13!!
